qa: upgrade only from N-2, N-1 releases

dparmar18 opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

dparmar18 commented:
Code works well and will update the YAML fragments dynamically; I had put this on testing but I'm not sure why pacific->reef fails, rest everything passes [0], [1]. The error is same across all the failed jobs(all the failed jobs being pacific->reef):
```
Command failed on smithi089 with status 124: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph pg dump --format=json'
```

The error looks unrelated to the changes I made, had a quick go-through the teuthology logs as well, so I changed reef to quincy and it passes [2]. Did we just found out a bug?

---
[0] http://pulpito.front.sepia.ceph.com/dparmar-2023-09-30_21:53:45-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/
[1] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-01_07:32:21-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/
[2] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-01_08:17:04-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/

dparmar18 commented:
FYI `nofs` changed passed http://pulpito.front.sepia.ceph.com/dparmar-2023-09-30_21:55:54-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/

dparmar18 commented:
This looks like a bug, while flushing all pg stats, it ran into an exception and then waited for all the pgs to be clean but it looks not all pgs could get cleaned. I see these lines in the teuthology logs:

```
2023-10-01T08:01:50.312 INFO:tasks.ceph.ceph_manager.ceph:need seq 128849018884 got 128849018882 for osd.1
2023-10-01T08:01:50.338 INFO:tasks.ceph:ignoring flush pg stats error, probably testing upgrade: timed out waiting for mon to be updated with osd.3: 137438953474 < 137438953475
2023-10-01T08:01:50.339 INFO:tasks.ceph.ceph_manager.ceph:waiting for clean
```

and then fails:
```
2023-10-01T08:01:51.815 INFO:tasks.ceph.ceph_manager.ceph:need seq 124554051588 got 124554051586 for osd.0
2023-10-01T08:03:50.390 DEBUG:teuthology.orchestra.run:got remote process result: 124
```

These lines are existent in the pacific->reef runs while non-existent in the pacific->quincy run shared above

dparmar18 commented:
Ran all the `upgraded_client` tests(just like [0] but with pacific->quincy) and as expected [1] all passed. It looks like while upgrading from pacific to reef, the PGs are not getting cleaned

---
[0] http://pulpito.front.sepia.ceph.com/dparmar-2023-09-30_21:53:45-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/
[1] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-03_07:52:28-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/

vshankar commented on qa/suites/fs/upgrade/upgraded_client/tasks/2-upgrade.yaml:
```diff
@@ -0,0 +1,63 @@
+              osd_release = "ceph osd require-osd-release pacific"
+            elseif tostring(branch) == "pacific" then
+              upgrade_branch = "reef"
+              osd_release = "ceph osd require-osd-release reef"
```
This is applied _after_ the upgrade, yes? I wonder if the failures you see with p->r upgrade have something to do with this.

dparmar18 replied:
it then had to fail with nautilus->pacific, also this is the same upgrade yaml found in any other suite

vshankar replied:
OK. I've not yet gone through the failed job yet. Will have a look.

vshankar commented:
@dparmar18 as we discussed - could you check if rados upgrade has the same issue you see (PGs not getting active).

dparmar18 commented:
> @dparmar18 as we discussed - could you check if rados upgrade has the same issue you see (PGs not getting active).

  i checked it yesterday; that suite is very small and has a starting point from quincy and upgrades to reef, so i followed your advice and sent out a detailed text on ceph-devel channel https://ceph-storage.slack.com/archives/C1HFZTW81/p1696429501726789

dparmar18 commented:
this means there was no coverage of p->r until now

vshankar commented:
> Ran all the `upgraded_client` tests(just like [0] but with pacific->quincy) and as expected [1] all passed. It looks like while upgrading from pacific to reef, the PGs are not getting cleaned
> 
> [0] http://pulpito.front.sepia.ceph.com/dparmar-2023-09-30_21:53:45-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/ [1] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-03_07:52:28-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/

I'll try to have a look today. In the meantime tagging @ljflores to check if this is seen elsewhere.

dparmar18 commented:
linking the detailed note i shared on upstream slack channel https://ceph-storage.slack.com/archives/C1HFZTW81/p1696429501726789

vshankar commented:
@dparmar18 I think the crash is related to https://github.com/ceph/ceph/pull/51041

dparmar18 commented:
> > Ran all the `upgraded_client` tests(just like [0] but with pacific->quincy) and as expected [1] all passed. It looks like while upgrading from pacific to reef, the PGs are not getting cleaned
> > [0] http://pulpito.front.sepia.ceph.com/dparmar-2023-09-30_21:53:45-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/ [1] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-03_07:52:28-fs:upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/
> 
> @dparmar18 btw the failed tests (p->r upgrade) has a core file:
> 
> ```
> file 1696112205.126953.core
> 1696112205.126953.core: ELF 64-bit LSB core file, x86-64, version 1 (SYSV), SVR4-style, from 'ceph-mgr -f --cluster ceph -i x', real uid: 0, effective uid: 0, real gid: 0, effective gid: 0, execfn: '/bin/ceph-mgr', platform: 'x86_64'
> ```

do you want me debug it using gdb and dump some useful info?

vshankar commented:
> do you want me debug it using gdb and dump some useful info?

I have everything required to debug.

vshankar commented:
> > do you want me debug it using gdb and dump some useful info?
> 
> I have everything required to debug.

I tested a fix for this - will push a PR.

vshankar commented:
@dparmar18 https://tracker.ceph.com/issues/63188

vshankar commented:
> @dparmar18 https://tracker.ceph.com/issues/63188

@dparmar18 Could you do a quick test with the fix for this change? You can use the build which has the fixes and use the qa suite change here :)

dparmar18 commented:
> > @dparmar18 https://tracker.ceph.com/issues/63188
> 
> @dparmar18 Could you do a quick test with the fix for this change? You can use the build which has the fixes and use the qa suite change here :)

sure, can you tell me the build name that has the fix?

dparmar18 commented:
> > > @dparmar18 https://tracker.ceph.com/issues/63188
> > 
> > 
> > @dparmar18 Could you do a quick test with the fix for this change? You can use the build which has the fixes and use the qa suite change here :)
> 
> sure, can you tell me the build name that has the fix?

running fs:upgrade suite using build `wip-vshankar-testing-20231018.065603` with my repo and suite branch `wip-62953` (this PR): http://pulpito.front.sepia.ceph.com/dparmar-2023-10-19_13:06:44-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/

vshankar commented:
> @vshankar http://pulpito.front.sepia.ceph.com/dparmar-2023-10-19_13:06:44-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/7432330/
> 
> ```
> Command failed on smithi060 with status 124: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph pg dump --format=json'
> ```
> 
> I.e. same failure as faced earlier.
> 
> also has the core dump http://qa-proxy.ceph.com/teuthology/dparmar-2023-10-19_13:06:44-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/7432330/remote/smithi060/coredump/

@dparmar18 - this will fail since the test upgrades to reef which doesn't have the fix yet. You'd need to test to upgrade to main from quincy-release.

I'll test it out.

dparmar18 commented:
> > @vshankar http://pulpito.front.sepia.ceph.com/dparmar-2023-10-19_13:06:44-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/7432330/
> > ```
> > Command failed on smithi060 with status 124: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph pg dump --format=json'
> > ```
> > 
> > 
> > I.e. same failure as faced earlier.
> > also has the core dump http://qa-proxy.ceph.com/teuthology/dparmar-2023-10-19_13:06:44-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/7432330/remote/smithi060/coredump/
> 
> @dparmar18 - this will fail since the test upgrades to reef which doesn't have the fix yet. You'd need to test to upgrade to main from quincy-release.
> 
> I'll test it out.

In this PR every YAML has a branch name mentioned while upgrading (such that it is not more than N-2) so upgrading to main is not possible at least with this PR, for that we will need to add a case. I can do it as a separate commit in this PR and once confirmed it passes, I'll drop it. What do you think?

vshankar commented:
> In this PR every YAML has a branch name mentioned(octopus, pacific, quincy, reef but not main) while upgrading (such that it is not more than N+2) so upgrading to main is not possible at least with this PR, for that we will need to add a case. I can do it as a separate commit in this PR and once confirmed it passes, I'll drop it. What do you think?

Never mind. I'm going to merge the fixes today and backport to reef. We can test it out then.

vshankar commented:
@dparmar18 https://github.com/ceph/ceph/pull/54146 is merged. This can be put to test again.

dparmar18 commented:
last push was just a rebase

dparmar18 commented:
@vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/

dparmar18 commented:
> @vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/

there are relevant failures

dparmar18 commented:
> > > @vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/
> > 
> > 
> > there are relevant failures
> 
> Never mind - install.upgrade is fetching
> 
> > 2023-10-31T11:23:01.059 INFO:teuthology.task.install:Ceph rpm upgrade from 16.2.14-98.g24209a7c.el8 to 18.2.0-1130.g1f1d869f
> 
> which doesn't have the fixes yet.

oops, when can we expect `18.2.0-1130.g1f1d869f` to have this?

dparmar18 commented:
> > > > > @vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/
> > > > 
> > > > 
> > > > there are relevant failures
> > > 
> > > 
> > > Never mind - install.upgrade is fetching
> > > > 2023-10-31T11:23:01.059 INFO:teuthology.task.install:Ceph rpm upgrade from 16.2.14-98.g24209a7c.el8 to 18.2.0-1130.g1f1d869f
> > > 
> > > 
> > > which doesn't have the fixes yet.
> > 
> > 
> > oops, when can we expect `18.2.0-1130.g1f1d869f` to have this?
> 
> Probably when we have a fresh build here - https://shaman.ceph.com/builds/ceph/reef/

We have `18.2.0-1152-g2ca2baeb` now. Worth giving it a try I guess

dparmar18 commented:
> > > > > > @vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/
> > > > > 
> > > > > 
> > > > > there are relevant failures
> > > > 
> > > > 
> > > > Never mind - install.upgrade is fetching
> > > > > 2023-10-31T11:23:01.059 INFO:teuthology.task.install:Ceph rpm upgrade from 16.2.14-98.g24209a7c.el8 to 18.2.0-1130.g1f1d869f
> > > > 
> > > > 
> > > > which doesn't have the fixes yet.
> > > 
> > > 
> > > oops, when can we expect `18.2.0-1130.g1f1d869f` to have this?
> > 
> > 
> > Probably when we have a fresh build here - https://shaman.ceph.com/builds/ceph/reef/
> 
> We have `18.2.0-1152-g2ca2baeb` now. Worth giving it a try I guess

`reef` in `ceph-ci` still doesn't have the commit https://github.com/ceph/ceph-ci/blob/reef/src/client/Client.cc

vshankar commented:
> > > > > > > @vshankar scheduled a run with your build `wip-vshankar-testing-20231018.065603` against this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-10-31_11:02:23-fs:upgrade-wip-vshankar-testing-20231018.065603-distro-default-smithi/
> > > > > > 
> > > > > > 
> > > > > > there are relevant failures
> > > > > 
> > > > > 
> > > > > Never mind - install.upgrade is fetching
> > > > > > 2023-10-31T11:23:01.059 INFO:teuthology.task.install:Ceph rpm upgrade from 16.2.14-98.g24209a7c.el8 to 18.2.0-1130.g1f1d869f
> > > > > 
> > > > > 
> > > > > which doesn't have the fixes yet.
> > > > 
> > > > 
> > > > oops, when can we expect `18.2.0-1130.g1f1d869f` to have this?
> > > 
> > > 
> > > Probably when we have a fresh build here - https://shaman.ceph.com/builds/ceph/reef/
> > 
> > 
> > We have `18.2.0-1152-g2ca2baeb` now. Worth giving it a try I guess
> 
> `reef` in `ceph-ci` still doesn't have the commit https://github.com/ceph/ceph-ci/blob/reef/src/client/Client.cc

Why would that be the case?

vshankar commented:
@dparmar18 https://pulpito.ceph.com/vshankar-2023-11-06_09:35:56-fs:upgrade-wip-vshankar-testing-20231106.073650-testing-default-smithi/

PRs in test - https://github.com/ceph/ceph/pulls?q=is%3Aopen+label%3Acephfs+label%3Awip-vshankar-testing

The (one) failure is the one you mentioned in slack:

```
2023-11-06T11:02:29.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:2023-11-06T11:02:29.348+0000 7faeb63d75c0 -1 init, newargv = 0x55aabe5e3c30 newargc=15
2023-11-06T11:02:29.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:ceph-fuse[139339]: starting ceph client
2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:terminate called after throwing an instance of 'ceph::buffer::v15_2_0::end_of_buffer'
2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:  what():  End of buffer [buffer:2]
2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:*** Caught signal (Aborted) **
2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: in thread 7fae8effd700 thread_name:ms_dispatch
2023-11-06T11:02:29.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: ceph version 18.0.0-7083-g85bdb240 (85bdb240f5cf2a758acc9b247951fbc8f5799bc1) reef (dev)
2023-11-06T11:02:29.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 1: /lib64/libpthread.so.0(+0x12cf0) [0x7faeb3012cf0]
2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 2: gsignal()
2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 3: abort()
2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 4: /lib64/libstdc++.so.6(+0x9009b) [0x7faeb149009b]
2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 5: /lib64/libstdc++.so.6(+0x9653c) [0x7faeb149653c]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 6: /lib64/libstdc++.so.6(+0x96597) [0x7faeb1496597]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 7: /lib64/libstdc++.so.6(+0x967f8) [0x7faeb14967f8]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 8: (ceph::buffer::v15_2_0::list::iterator_impl<true>::copy(unsigned int, char*)+0xa5) [0x7faeb489f475]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 9: (MDSMap::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&)+0xa54) [0x7faeb4ac8084]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 10: (Client::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&)+0x523) [0x55aabd727823]
2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 11: (Client::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x51d) [0x55aabd72881d]
2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 12: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x478) [0x7faeb47419c8]
2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 13: (DispatchQueue::entry()+0x51f) [0x7faeb473eb4f]
2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 14: (DispatchQueue::DispatchThread::entry()+0x11) [0x7faeb480c3d1]
2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 15: /lib64/libpthread.so.0(+0x81ca) [0x7faeb30081ca]
2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 16: clone()
```

dparmar18 commented:
`2-upgrade.yaml` doesn't make sense in `fs/upgrade/upgraded_client` sub-suite since this was aimed to test new client against old clusters, therefore the last commit made removed it along with few changes:

- moved `0-clients/` from `tasks/3-workload/new_ops/`  to `tasks/`  and renamed it to `2-clients/`  (i.e. replacing `2-upgrade.yaml`)
- since `new_ops/`  and `stress_tests/`  now share the common upgrade yaml, moved the tests yamls(in `stress_tests/`) directly within `3-workload/stress_tests/`
- renamed `1-client-sanity.yaml`  in `new_ops/`  to `newops.yaml`  (however better option was remove the `new_ops` dir and directly have the `newops.yaml`  but opted for this since we might want to add new yamls in the dir?)

vshankar commented:
> @dparmar18 https://pulpito.ceph.com/vshankar-2023-11-06_09:35:56-fs:upgrade-wip-vshankar-testing-20231106.073650-testing-default-smithi/
> 
> PRs in test - https://github.com/ceph/ceph/pulls?q=is%3Aopen+label%3Acephfs+label%3Awip-vshankar-testing
> 
> The (one) failure is the one you mentioned in slack:
> 
> ```
> 2023-11-06T11:02:29.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:2023-11-06T11:02:29.348+0000 7faeb63d75c0 -1 init, newargv = 0x55aabe5e3c30 newargc=15
> 2023-11-06T11:02:29.349 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:ceph-fuse[139339]: starting ceph client
> 2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:terminate called after throwing an instance of 'ceph::buffer::v15_2_0::end_of_buffer'
> 2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:  what():  End of buffer [buffer:2]
> 2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr:*** Caught signal (Aborted) **
> 2023-11-06T11:02:29.353 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: in thread 7fae8effd700 thread_name:ms_dispatch
> 2023-11-06T11:02:29.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: ceph version 18.0.0-7083-g85bdb240 (85bdb240f5cf2a758acc9b247951fbc8f5799bc1) reef (dev)
> 2023-11-06T11:02:29.354 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 1: /lib64/libpthread.so.0(+0x12cf0) [0x7faeb3012cf0]
> 2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 2: gsignal()
> 2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 3: abort()
> 2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 4: /lib64/libstdc++.so.6(+0x9009b) [0x7faeb149009b]
> 2023-11-06T11:02:29.358 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 5: /lib64/libstdc++.so.6(+0x9653c) [0x7faeb149653c]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 6: /lib64/libstdc++.so.6(+0x96597) [0x7faeb1496597]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 7: /lib64/libstdc++.so.6(+0x967f8) [0x7faeb14967f8]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 8: (ceph::buffer::v15_2_0::list::iterator_impl<true>::copy(unsigned int, char*)+0xa5) [0x7faeb489f475]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 9: (MDSMap::decode(ceph::buffer::v15_2_0::list::iterator_impl<true>&)+0xa54) [0x7faeb4ac8084]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 10: (Client::handle_mds_map(boost::intrusive_ptr<MMDSMap const> const&)+0x523) [0x55aabd727823]
> 2023-11-06T11:02:29.359 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 11: (Client::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x51d) [0x55aabd72881d]
> 2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 12: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x478) [0x7faeb47419c8]
> 2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 13: (DispatchQueue::entry()+0x51f) [0x7faeb473eb4f]
> 2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 14: (DispatchQueue::DispatchThread::entry()+0x11) [0x7faeb480c3d1]
> 2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 15: /lib64/libpthread.so.0(+0x81ca) [0x7faeb30081ca]
> 2023-11-06T11:02:29.360 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi150.stderr: 16: clone()
> ```

@dparmar18 This crash is due to absence of PR #53340 in reef. The MDS will encode the following two fields in the MDSMap with version=17:

```
  encode(max_xattr_size, bl);
  encode(bal_rank_mask, bl);
```

But the client running reef version, will decode like:

```
  if (ev >= 17) {
    decode(bal_rank_mask, p);
  }
```

and then the final `DECODE_FINISH(p);` will result in this crash. See discussion here - https://github.com/ceph/ceph/pull/46357#issuecomment-1293556227

dparmar18 commented:
things went fine https://pulpito.ceph.com/dparmar-2023-11-07_06:18:43-fs:upgrade-main-distro-default-smithi/

vshankar commented:
> things went fine https://pulpito.ceph.com/dparmar-2023-11-07_06:18:43-fs:upgrade-main-distro-default-smithi/

Right, but we found an issue which eventually would have been resolved :)

dparmar18 commented:
Short summary of the changes made in this PR: 

- upgrade of the cluster/client is limited to at max N+2 from the base version i.e. the upgrades in upgrade suite as now range bounded
- removal of `2-upgrade.yaml` from `upgrade/upgraded_client/` cause the aim of the sub-suite (at time of writting https://github.com/ceph/ceph/pull/48280 ) is to test newer client against an old cluster.
- some refactoring to remove redundant upgrade yaml found in `upgrade/upgraded_client/tasks/3-workload/stress_tests` and `upgrade/upgraded_client/tasks/3-workload/new_ops` (now both will make use of single global client upgrade yaml)
- renaming of yaml in `upgrade/upgraded_client/tasks/3-workload/new_ops` to `new_ops.yaml`

dparmar18 commented:
dir `qa/suites/fs/upgrade/upgraded_client/`

tree before this patch:
```
├── %
├── bluestore-bitmap.yaml -> ../../../../cephfs/objectstore-ec/bluestore-bitmap.yaml
├── centos_8.stream.yaml -> .qa/distros/all/centos_8.stream.yaml
├── clusters
│   ├── %
│   └── 1-mds-1-client-micro.yaml -> .qa/cephfs/clusters/1-mds-1-client-micro.yaml
├── conf -> .qa/cephfs/conf/
├── overrides
│   ├── %
│   ├── ignorelist_health.yaml -> .qa/cephfs/overrides/ignorelist_health.yaml
│   ├── ignorelist_wrongly_marked_down.yaml -> .qa/cephfs/overrides/ignorelist_wrongly_marked_down.yaml
│   └── pg-warn.yaml
└── tasks
    ├── %
    ├── 0-from
    │   ├── nautilus.yaml
    │   └── pacific.yaml
    ├── 1-mount
    │   └── mount -> .qa/cephfs/mount/
    └── 2-workload
        ├── new_ops
        │   ├── %
        │   ├── 0-clients
        │   │   ├── fuse-upgrade.yaml
        │   │   └── kclient.yaml
        │   └── 1-client-sanity.yaml
        └── stress_tests
            ├── %
            ├── 0-client-upgrade.yaml
            └── 1-tests
                ├── blogbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/blogbench.yaml
                ├── dbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/dbench.yaml
                ├── fsstress.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/fsstress.yaml
                ├── iozone.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/iozone.yaml
                └── kernel_untar_build.yaml -> .qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml
```
---

tree after this patch:
```
├── %
├── bluestore-bitmap.yaml -> ../../../../cephfs/objectstore-ec/bluestore-bitmap.yaml
├── branch
│   ├── nautilus.yaml
│   └── pacific.yaml
├── centos_8.stream.yaml -> .qa/distros/all/centos_8.stream.yaml
├── clusters
│   ├── %
│   └── 1-mds-1-client-micro.yaml -> .qa/cephfs/clusters/1-mds-1-client-micro.yaml
├── conf -> .qa/cephfs/conf/
├── overrides
│   ├── %
│   ├── ignorelist_health.yaml -> .qa/cephfs/overrides/ignorelist_health.yaml
│   ├── ignorelist_wrongly_marked_down.yaml -> .qa/cephfs/overrides/ignorelist_wrongly_marked_down.yaml
│   └── pg-warn.yaml
└── tasks
    ├── %
    ├── 0-install.yaml
    ├── 1-mount
    │   └── mount -> .qa/cephfs/mount/
    ├── 2-clients
    │   ├── fuse-upgrade.yaml
    │   └── kclient.yaml
    └── 3-workload
        ├── new_ops
        │   ├── %
        │   └── newops.yaml
        └── stress_tests
            ├── blogbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/blogbench.yaml
            ├── dbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/dbench.yaml
            ├── fsstress.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/fsstress.yaml
            ├── iozone.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/iozone.yaml
            └── kernel_untar_build.yaml -> .qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml
```

vshankar commented:
> dir `qa/suites/fs/upgrade/upgraded_client/`
> 
> tree before this patch:
> 
> ```
> ├── %
> ├── bluestore-bitmap.yaml -> ../../../../cephfs/objectstore-ec/bluestore-bitmap.yaml
> ├── centos_8.stream.yaml -> .qa/distros/all/centos_8.stream.yaml
> ├── clusters
> │   ├── %
> │   └── 1-mds-1-client-micro.yaml -> .qa/cephfs/clusters/1-mds-1-client-micro.yaml
> ├── conf -> .qa/cephfs/conf/
> ├── overrides
> │   ├── %
> │   ├── ignorelist_health.yaml -> .qa/cephfs/overrides/ignorelist_health.yaml
> │   ├── ignorelist_wrongly_marked_down.yaml -> .qa/cephfs/overrides/ignorelist_wrongly_marked_down.yaml
> │   └── pg-warn.yaml
> └── tasks
>     ├── %
>     ├── 0-from
>     │   ├── nautilus.yaml
>     │   └── pacific.yaml
>     ├── 1-mount
>     │   └── mount -> .qa/cephfs/mount/
>     └── 2-workload
>         ├── new_ops
>         │   ├── %
>         │   ├── 0-clients
>         │   │   ├── fuse-upgrade.yaml
>         │   │   └── kclient.yaml
>         │   └── 1-client-sanity.yaml
>         └── stress_tests
>             ├── %
>             ├── 0-client-upgrade.yaml
>             └── 1-tests
>                 ├── blogbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/blogbench.yaml
>                 ├── dbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/dbench.yaml
>                 ├── fsstress.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/fsstress.yaml
>                 ├── iozone.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/iozone.yaml
>                 └── kernel_untar_build.yaml -> .qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml
> ```
> 
> tree after this patch:
> 
> ```
> ├── %
> ├── bluestore-bitmap.yaml -> ../../../../cephfs/objectstore-ec/bluestore-bitmap.yaml
> ├── branch
> │   ├── nautilus.yaml
> │   └── pacific.yaml
> ├── centos_8.stream.yaml -> .qa/distros/all/centos_8.stream.yaml
> ├── clusters
> │   ├── %
> │   └── 1-mds-1-client-micro.yaml -> .qa/cephfs/clusters/1-mds-1-client-micro.yaml
> ├── conf -> .qa/cephfs/conf/
> ├── overrides
> │   ├── %
> │   ├── ignorelist_health.yaml -> .qa/cephfs/overrides/ignorelist_health.yaml
> │   ├── ignorelist_wrongly_marked_down.yaml -> .qa/cephfs/overrides/ignorelist_wrongly_marked_down.yaml
> │   └── pg-warn.yaml
> └── tasks
>     ├── %
>     ├── 0-install.yaml
>     ├── 1-mount
>     │   └── mount -> .qa/cephfs/mount/
>     ├── 2-clients
>     │   ├── fuse-upgrade.yaml
>     │   └── kclient.yaml
>     └── 3-workload
>         ├── new_ops
>         │   ├── %
>         │   └── newops.yaml
>         └── stress_tests
>             ├── blogbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/blogbench.yaml
>             ├── dbench.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/dbench.yaml
>             ├── fsstress.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/fsstress.yaml
>             ├── iozone.yaml -> .qa/suites/fs/workload/tasks/5-workunit/suites/iozone.yaml
>             └── kernel_untar_build.yaml -> .qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml
> ```

@dparmar18 Could you explain (in the commit message) why there is a "topology" change in the way the yamls are structured? IIUC, there are changes that auto-select branch names, but it would help to be clear in the commit message.

dparmar18 commented:
> @dparmar18 Could you explain (in the commit message) why there is a "topology" change in the way the yamls are structured? IIUC, there are changes that auto-select branch names, but it would help to be clear in the commit message.

the cluster should never be upgraded in `upgrade/upgraded_client/` since the aim of the sub-suite is to test newer client against an old cluster therefore removed `2-upgrade.yaml` from the dir.

`upgrade/upgraded_client/tasks/3-workload/stress_tests` and `upgrade/upgraded_client/tasks/3-workload/new_ops` had a upgrade yaml which can be avoided and a single upgrade yaml can be used, so there is no `upgraded_client/tasks/3-workload/new_ops/0-clients` and `upgrade/upgraded_client/tasks/3-workload/stress_tests/0-client-upgrade.yaml` and thus that complication has been removed, now since there is no `0-client-upgrade.yaml`, `upgrade/upgraded_client/tasks/3-workload/stress_tests/1-tests`'s tests have been directly under `upgrade/upgraded_client/tasks/3-workload/stress_tests`

the new ops yaml in` upgrade/upgraded_client/tasks/3-workload/new_ops` was named `1-client-sanity.yaml` which is renamed to `new_ops.yaml`

dparmar18 commented:
@vshankar the explanation is there in the commit already, do let me know if its not enough

vshankar commented:
> @vshankar the explanation is there in the commit already, do let me know if its not enough

I see you updated the commit message. Nice clear description now - appreciate it @dparmar18

vshankar commented:
Putting this to test since the fix has been merged.

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#29-Jan-2024
